spmd/compiler/fusion.py

```diff
@@ -162,22 +214,448 @@ def _scan_graph_for_fusion_elements(
+        updated_dtype,
+        requires_grad,
+        stride,
+        updated_memory_format,
+        is_quantized,
+        qparams,
+    )
+
+    # update meta with new TensorMetadata
+    saved_meta = node.meta.get("tensor_meta")
+    node.meta["tensor_meta"] = new_metadata
+
+    return new_metadata
+
+
+def _finalize_output_node(
+    gi: GraphInfo,
+    gm: fx.GraphModule,
+    fe_list: List[FusionElement],
+    start: int,
+    stop: int,
+    new_output_args: List[fx.Node],
+) -> None:
+    """reworks output node args to original grad tensors, replacing the wait_comms
+    we update a copy of the output args, then finalized after all fusion is done."""
+
+    # output_node = gi.output
+    # new_output_args = []
+
+    # curr_output_args = list(gi.output.args[0])
+    replacement_mapping: Dict[fx.Node, fx.Node] = {}
+
+    # map out all updated nodes in our list
+    for item in fe_list:
+        grad_node = item.grad_tensor_node
+        wait_node = item.wait_node
+        replacement_mapping[wait_node] = grad_node  # type: ignore
+
+    # we have fused a subset, only update that subset within the larger output node args
+    # TODO - this assumes that all gradient tensors are comm handled.
+    for i in range(len(fe_list)):
+        index = start + i
+        curr_node = new_output_args[index]
+
+        if curr_node is not None:
+            assert curr_node.name.startswith(
+                "wait"
+            ), f"Non comm gradient output tensor incorrectly handled...needs fix. {new_output_args[start+i]}"
+            new_output_args[start + i] = replacement_mapping[curr_node]
+
+    # _debug(f"537 - updated output args = {new_output_args}\n")
+
+
+def _determine_peak_memory(gi: GraphInfo, fusion_policy: int) -> int:
+    """
+    scans fe list to determine max memory required across all fusion instances.
+    this result is used to allocate the global buffer for fusion, where we
+    re-use a global buffer to avoid repeated allocations per fusion.
+    """
+    peak_memory = 0  # currently measured in numel
+    curr_memory = 0
+    fast_index = 0
+    for i, item in enumerate(gi.fe_list):  # type: ignore
+        fast_index += 1
+        curr_memory += item.size  # type: ignore
+
+        if fast_index == fusion_policy:
+            peak_memory = max(peak_memory, curr_memory)
+            fast_index = 0
+            curr_memory = 0
+
+    # _debug(f"574, peak memory determined to be {peak_memory}")
+    gi.peak_memory_required = peak_memory
+
+    return peak_memory
+
+
+def run_comm_fusion(gm: fx.GraphModule) -> fx.GraphModule:
     """main entry into remapping graph for all_reduce fusion"""
 
-    result = False
+    # result = False  TODO - do we return a success code or the graph,
+    # since graph is modified directly in place?
+
+    # first recompile to make sure we have coherent graph
+    gm.recompile()
 
     # get our main graph info
-    graph_info = GraphInfo()
-    graph_info.update_info(gm)
+    gi = GraphInfo()
```

Comment:
nit: lets rename to graph_info for readability

Reply:
updated to graph_info!